File: src/main/java/no/nav/syfo/services/ws/CallIdHeader.java
package no.nav.syfo.services.ws;

import org.apache.cxf.binding.soap.SoapHeader;
import org.apache.cxf.binding.soap.SoapMessage;
import org.apache.cxf.interceptor.Fault;
import org.apache.cxf.jaxb.JAXBDataBinding;
import org.apache.cxf.message.Message;
import org.apache.cxf.phase.AbstractPhaseInterceptor;
import org.apache.cxf.phase.Phase;
import org.slf4j.Logger;
import org.slf4j.MDC;

import javax.xml.bind.JAXBException;
import javax.xml.namespace.QName;

import static java.util.Optional.ofNullable;
import static java.util.UUID.randomUUID;
import static org.slf4j.LoggerFactory.getLogger;

public class CallIdHeader extends AbstractPhaseInterceptor<Message> {

    private static final Logger log = getLogger(CallIdHeader.class);

    public CallIdHeader() {
        super(Phase.PRE_STREAM);
    }

    @Override
    public void handleMessage(Message message) throws Fault {
        try {
            QName qName = new QName("uri:no.nav.applikasjonsrammeverk", "callId");
            SoapHeader header = new SoapHeader(qName, callId(), new JAXBDataBinding(String.class));
            ((SoapMessage) message).getHeaders().add(header);
        } catch (JAXBException ex) {
            log.warn("Error while setting CallId header", ex);
        }
    }

    private String callId() {
        return ofNullable(MDC.get("Nav-Callid"))
                .orElseGet(randomUUID()::toString);
    }

}

File: src/main/kotlin/no/nav/syfo/consumer/util/ws/CallIdHeader.kt
package no.nav.syfo.consumer.util.ws

import org.apache.cxf.binding.soap.SoapHeader
import org.apache.cxf.binding.soap.SoapMessage
import org.apache.cxf.interceptor.Fault
import org.apache.cxf.jaxb.JAXBDataBinding
import org.apache.cxf.message.Message
import org.apache.cxf.phase.AbstractPhaseInterceptor
import org.apache.cxf.phase.Phase
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import java.util.*
import javax.xml.bind.JAXBException
import javax.xml.namespace.QName

class CallIdHeader : AbstractPhaseInterceptor<Message>(Phase.PRE_STREAM) {
    @Throws(Fault::class)
    override fun handleMessage(message: Message) {
        try {
            val qName = QName("uri:no.nav.applikasjonsrammeverk", "callId")
            val header = SoapHeader(qName, callId(), JAXBDataBinding(String::class.java))
            (message as SoapMessage).headers.add(header)
        } catch (ex: JAXBException) {
            log.warn("Error while setting CallId header", ex)
        }
    }

    private fun callId(): String {
        return Optional.ofNullable(MDC.get("Nav-Callid"))
            .orElseGet { UUID.randomUUID().toString() }
    }

    companion object {
        private val log = LoggerFactory.getLogger(CallIdHeader::class.java)
    }
}
